lib/elixir/lib/map_set.ex

```diff
@@ -347,6 +347,8 @@ defmodule MapSet do
 
   """
   @spec union(t(val1), t(val2)) :: t(val1 | val2) when val1: value, val2: value
+  def union(map_set1, map_set2)
```

Comment:
ok, I found the offending commit doing a git bisect, 
https://github.com/elixir-lang/elixir/commit/bc9df398759769a7656589b51e32809071e8df7c#diff-0a5412a70e42a886648e41d22f7d17aeR350

the good thing is that the fix that I have for https://github.com/elixir-lang/elixir/pull/6195 solves this.
but instead of doing:
`def union(%MapSet{map: map1, version: version} = map_set, %MapSet{map: map2, version: version}) do`
we will have to do
`def union(%MapSet{map: map1, version: version} = map_set1, %MapSet{map: map2, version: version} = _map_set2) do`
otherwise the signature will be `union(map_set, map_set)`
and we want `union(map_set1, map_set2)`
So i think it is better to introduce my change.